Scripts/Runtime/Entities/ManagedReference/ManagedReferenceStore.cs

```diff
@@ -0,0 +1,72 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Anvil.Unity.DOTS.Data
+{
+    /// <summary>
+    /// A static store of all <see cref="IComponentReferencable"/> instances. Used to get instances by hash.
+    /// Used by <see cref="ManagedReference{T}"/>.
+    /// </summary>
+    public static class ManagedReferenceStore
+    {
+        private static readonly Dictionary<int, IComponentReferencable> m_instances = new Dictionary<int, IComponentReferencable>();
+
+#if UNITY_EDITOR
+        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.SubsystemRegistration)]
+        private static void Init()
+        {
+            m_instances.Clear();
+        }
+#endif
+
+        /// <summary>
+        /// Registers a managed instance so it can be resolved by a <see cref="ManagedReference{T}"/>.
+        /// This is usually called when the instance is created.
+        /// </summary>
+        /// <param name="instance">The instance to register.</param>
+        public static void RegisterReference(IComponentReferencable instance)
+        {
+            Debug.Assert(instance.GetType().IsClass, $"Implementations of {nameof(IComponentReferencable)} must be classes");
+            m_instances.Add(instance.GetHashCode(), instance);
```

Comment:
While it is unlikely, it's possible that two separate instances could generate the same hash code and we'll get a collision. 

This will throw an exception here which is probably fine for alerting us that we now need to deal with it. But just wanted to raise it in case we want to do some sort of assert check before. 

Unity's TypeManager goes WAY into depth with detecting and dealing with Hash Collisions which I don't think we need to do, but if we discover that we're getting hash collisions we might need to handle modifying the hash code.

Reply:
Yea, I just looked into this for a different reason. Checking if GetHashCode() would ever return 0 to see if I could easily detect a `default(ManagedRef<T>)` and found this:
https://docs.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-6.0#remarks

Soooo, I need to rethink hashcodes. I think if I mix in an ID provider I'll be fine but this is why I had a separate GetHash method to make this change easy!

Reply:
Reworked. See next PR post for details.